src/Cryptol/Parser/ParserUtils.hs

```diff
@@ -295,21 +300,51 @@ mkEApp es@(eLast : _) = at (eFirst,eLast) $ foldl EApp f xs
      becomes
      [ f, x ` { a = 2 }, y ]
   -}
-  cvtTypeParams e [] = [e]
+  cvtTypeParams e [] = pure (e :| [])
   cvtTypeParams e (p : ps) =
-    case toTypeParam p of
-      Just fs -> cvtTypeParams (EAppT e fs) ps
-      Nothing -> e : cvtTypeParams p ps
-
-  toTypeParam e =
-    case dropLoc e of
-      ETypeVal t -> case dropLoc t of
-                      TTyApp fs -> Just (map mkTypeInst fs)
-                      _         -> Nothing
+    case toTypeParam p Nothing of
+      Nothing -> NE.cons e <$> cvtTypeParams p ps
+
```

Comment:
It might be good to add a comment example to illustrate what case are we handling here. Something like
`` [ x, `{ a = 2 }.y  ] ~> [ (x`{a=2}).y ]``